File: src/main/java/ca/alexgirard/harpoonIJ/HarpoonDialog.java
package ca.alexgirard.harpoonIJ;

import com.intellij.ide.plugins.PluginManagerCore;
import com.intellij.openapi.editor.LogicalPosition;
import com.intellij.openapi.extensions.PluginId;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.EditorTextField;
import com.maddyhome.idea.vim.api.VimInjectorKt;
import com.maddyhome.idea.vim.command.MappingMode;
import com.maddyhome.idea.vim.key.MappingOwner;
import org.jetbrains.annotations.Nullable;

import com.maddyhome.idea.vim.KeyHandler;
import com.maddyhome.idea.vim.newapi.IjVimEditorKt;

import javax.swing.*;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.awt.event.KeyEvent;

public class HarpoonDialog extends DialogWrapper {

    public EditorTextField editorTextField;
    private final String text;
    public int SelectedIndex = -1;
    private static boolean enterRemapped = false;
    private boolean normalModeForcedAlready = false;

    public void Next() {

        var editor = editorTextField.getEditor();
        if (editor != null) {
            editor.getCaretModel().moveCaretRelatively(0, 1, false, false, false);
        }
    }

    public void Previous() {

        var editor = editorTextField.getEditor();
        if (editor != null) {
            editor.getCaretModel().moveCaretRelatively(0, -1, false, false, false);
        }
    }

    public void Ok() {

        var editor = editorTextField.getEditor();
        if (editor != null) {
            SelectedIndex = editor.getCaretModel().getLogicalPosition().line;
        }
        doOKAction();
    }

    protected HarpoonDialog(String inputText) {
        super(true);
        AppSettingsState settings = AppSettingsState.getInstance();
        setSize(settings.dialogWidth, settings.dialogHeight);
        setTitle("Harpoon");
        var vimPlugin = PluginManagerCore.getPlugin(PluginId.getId("IdeaVIM"));
        var vimEnabled = vimPlugin != null && vimPlugin.isEnabled();
        if (vimEnabled) {
            setEnterRemap(settings);
        }
        text = inputText;
        init();
    }

    //TODO move this to app startup or something 
    private void setEnterRemap(AppSettingsState settings) {
        if (enterRemapped || !settings.enterRemap) return;
        // i had 2 <cr> here before and i can't remember why
        var keys = VimInjectorKt.injector.getParser().parseKeys(":action SelectHarpoonItem<cr>");
        var keyGroup = VimInjectorKt.injector.getKeyGroup();
        keyGroup.putKeyMapping(MappingMode.NVO,
                VimInjectorKt.injector.getParser().parseKeys("<cr>")
                , MappingOwner.Plugin.Companion.get("HarpoonIj"), keys, false);
        enterRemapped = true;
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return editorTextField;
    }

    @Override
    protected @Nullable JComponent createCenterPanel() {
        AppSettingsState appSettings = AppSettingsState.getInstance();
        editorTextField = new EditorTextField(text);
        editorTextField.setOneLineMode(false);
        editorTextField.addSettingsProvider(editor -> {
            editor.setFontSize(appSettings.dialogFontSize);
            editor.setInsertMode(true);
            editor.getCaretModel().moveToLogicalPosition(new LogicalPosition(0, 0));
            var settings = editor.getSettings();
            settings.setLineNumbersShown(true);
        });


        editorTextField.addFocusListener(new FocusListener() {
            public void focusGained(FocusEvent e) {
                if (normalModeForcedAlready) return;
                var editor = editorTextField.getEditor();
                if (editor == null) return;
                var vim = IjVimEditorKt.getVim(editor);
                if (!vim.getInsertMode()) return;
                var context = VimInjectorKt.injector.getExecutionContextManager().getEditorExecutionContext(vim);
                KeyHandler.getInstance().handleKey(vim, KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), context, KeyHandler.getInstance().getKeyHandlerState());
                normalModeForcedAlready = true;
                if (editor != null) {
                    editor.getCaretModel().moveCaretRelatively(0, 1, false, false, false);
                }
            }

            public void focusLost(FocusEvent e) {
            }
        });
        return editorTextField;
    }


    @Override
    protected JComponent createSouthPanel() {
        return null;
    }
}

File: src/main/kotlin/ca/alexgirard/harpoonIJ/HarpoonDialog.kt
package ca.alexgirard.harpoonIJ

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.ex.EditorEx
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.EditorTextField
import com.maddyhome.idea.vim.VimPlugin
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.command.OperatorArguments
import com.maddyhome.idea.vim.command.VimStateMachine
import java.awt.event.FocusEvent
import java.awt.event.FocusListener
import javax.swing.JComponent


//import com.maddyhome.idea.vim.VimPlugin;
class HarpoonDialog(inputText: String) : DialogWrapper(true) {
    lateinit var editorTextField: EditorTextField;
    private val text: String
    private var insertModeForced = false;
    private val placeholder = "Nothing set yet :(";

    init {
        val settings: AppSettingsState = AppSettingsState.instance
        setSize(settings.dialogWidth, settings.dialogHeight)
        title = "Harpoon"
        text = inputText
        init()
    }

    override fun getPreferredFocusedComponent(): JComponent? {
        return editorTextField
    }

    override fun createCenterPanel(): JComponent? {
        val appSettings: AppSettingsState = AppSettingsState.instance
        editorTextField = EditorTextField(text)
        editorTextField.setOneLineMode(false)
        editorTextField.addSettingsProvider { editor: EditorEx ->
            editor.setFontSize(appSettings.dialogFontSize)
            val settings = editor.settings
            settings.isLineNumbersShown = true
        }

//        editorTextField.addKeyListener(object : KeyListener{
//            override fun keyTyped(e: KeyEvent?) {
//                TODO("Not yet implemented")
//            }
//
//            override fun keyPressed(e: KeyEvent?) {
//                TODO("Not yet implemented")
//            }
//
//            override fun keyReleased(e: KeyEvent?) {
//                TODO("Not yet implemented")
//            }
//
//        } );
        editorTextField.addFocusListener(object : FocusListener {
            
            override fun focusGained(e: FocusEvent?) {
                
                if (insertModeForced) return;
                for (item: VimEditor in VimPlugin.getEditor().localEditors()) {
//                    if (item.getVirtualFile() == null && item.mode == VimStateMachine.Mode.INSERT) {
//                        insertModeForced = true;
//                        item.exitInsertMode(
//                            injector.executionContextManager.onEditor(item),
//                            OperatorArguments(isOperatorPending = false, count0 = 1, mode = VimStateMachine.Mode.VISUAL, subMode = VimStateMachine.SubMode.VISUAL_BLOCK)
//                        );
                        item.exitInsertMode(
                            injector.executionContextManager.onEditor(item),
                            OperatorArguments(isOperatorPending = false, count0 = 1, mode = VimStateMachine.Mode.COMMAND, subMode = VimStateMachine.SubMode.NONE)
                        );
                        insertModeForced = false;
//                    }
                    
                    
                }
            }

            //
            override fun focusLost(e: FocusEvent?) {
//
//                for (item: VimEditor in VimPlugin.getEditor().localEditors()) {
//                    if (item.insertMode && item.getVirtualFile() == null && item.mode == VimStateMachine.Mode.INSERT) {
//                        insertModeForced = true;
//                        item.exitInsertMode(
//                            injector.executionContextManager.onEditor(item),
//                            OperatorArguments(
//                                isOperatorPending = false,
//                                count0 = 1,
//                                mode = VimStateMachine.Mode.INSERT,
//                                subMode = VimStateMachine.SubMode.NONE
//                            )
//                        );
//                        item.exitInsertMode(
//                            injector.executionContextManager.onEditor(item),
//                            OperatorArguments(
//                                isOperatorPending = false,
//                                count0 = 1,
//                                mode = VimStateMachine.Mode.COMMAND,
//                                subMode = VimStateMachine.SubMode.NONE
//                            )
//                        );
//                    }
                }
//                for (item: VimEditor in VimPlugin.getEditor().localEditors()) {
////                    item.insertMode = false;
//                    editorTextField.text = editorTextField.text + item.getPath();
//                }
//            }
        })
//            override fun focusGained(e: FocusEvent) {
//
//                val editor = (VimEditrVimPlugin.getEditor();
//                editorTextField.text+="aa";
//                for (item: VimEditor in editor.localEditors()) {
//                    item.insertMode = false;
//                    editorTextField.text+=item.getPath()
//                }
//            }
//
//            override fun focusLost(e: FocusEvent) {
////                val editor = VimPlugin.getEditor();
////                for (item: VimEditor1 in editor.localEditors()) {
////                    item.insertMode = false;
////                    editorTextField.text+=item.getPath()
////                }
//            }
//        })
        //        if (logger.isDebugEnabled())
//            logger.debug("woweee");
//        for (var editor : VimPlugin.().localEditors()) {
//            string.append(editor.getPath());
//            string.append("\n");
//        }
//        editorTextField.setText(string.toString());
        return editorTextField
    }

    override fun createSouthPanel(): JComponent? {
        return null
    }

    companion object {
        private val logger = Logger.getInstance(HarpoonDialog::class.java.getName())
    }
}
